magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,516 @@
+      if score_part_index >= len(self.score_parts):
+        score_part = ScorePart()
+        self.score_parts.append(score_part)
+
+      part = Part(child, self.score_parts[score_part_index])
+      self.parts.append(part)
+      score_part_index = score_part_index + 1
+      if current_time_position > self.total_time:
+        self.total_time = current_time_position
+
+  def getTimeSignatures(self):
+    time_signatures = []
+    for part in self.parts:
+      for measure in part.measures:
+        if measure.time_signature is not None:
+          if not measure.time_signature in time_signatures:
+            # Prevent duplicate time signatures
+            time_signatures.append(measure.time_signature)
+
+    return time_signatures
+
+  def getKeySignatures(self):
+    key_signatures = []
+    for part in self.parts:
+      for measure in part.measures:
+        if measure.key_signature is not None:
+          if not measure.key_signature in key_signatures:
+            # Prevent duplicate key signatures
+            key_signatures.append(measure.key_signature)
+
+    if len(key_signatures) == 0:
+      # If there are no key signatures, add C major at the beginning
+      key_signature = KeySignature()
+      key_signature.time_position = 0
+      key_signatures.append(key_signature)
+
+    return key_signatures
+
+  def getTempos(self):
+    tempos = []
+    for part in self.parts:
+      for measure in part.measures:
+        for tempo in measure.tempos:
+          tempos.append(tempo)
+
+    # If no tempos, add a default of 120 at beginning
+    if len(tempos) == 0:
+        tempo = Tempo()
+        tempo.qpm = current_tempo
+        tempo.time_position = 0
+        tempos.append(tempo)
+
+    return tempos
+
+
+class ScorePart:
+  def __init__(self, xml_score_part = None):
+    self.xml_score_part = xml_score_part
+    self.part_name = ""
+    self.midi_channel = 1
+    self.midi_program = 1
+    if xml_score_part != None:
+      self.parse()
+
+  def parse(self):
+    global current_midi_channel
+
+    if self.xml_score_part.find("part-name") != None:
+      self.part_name = self.xml_score_part.find("part-name").text
+
+    xml_midi_instrument = self.xml_score_part.find("midi-instrument")
+    if xml_midi_instrument != None and \
+      xml_midi_instrument.find("midi-channel") != None and \
+      xml_midi_instrument.find("midi-program") != None:
+      self.midi_channel = int(xml_midi_instrument.find("midi-channel").text)
+      self.midi_program = int(xml_midi_instrument.find("midi-program").text)
+    else:
+      # If no MIDI info, increment MIDI channel and use default program
+      self.midi_channel = current_midi_channel + 1
+      current_midi_channel = self.midi_channel
+      self.midi_program = 1
+
+  def __str__(self):
+    score_str = "ScorePart: " + self.part_name
+    score_str += ", Channel: " + str(self.midi_channel)
+    score_str += ", Program: " + str(self.midi_program)
+    return score_str
+
+
+class Part:
```

Comment:
All of these classes should have a docstring describing what information they contain.